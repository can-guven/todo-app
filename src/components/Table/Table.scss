.table {
    width: 100%;
    border-collapse: collapse;
    border-spacing: 0;
    border: 1px solid #e0e0e0;
    font-size: 14px;
    line-height: 1.5;
    color: #333;
    background-color: #fff;
    border-radius: 4px;
    overflow: hidden;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.1);
    margin-bottom: 20px;
    table-layout: fixed;
    &-head {
        background-color: #f5f5f5;
        &-cell {
            padding: 10px 15px;
            text-align: left;
            font-weight: 500;
            border-bottom: 1px solid #e0e0e0;
        }
    }
    &-body {
        &-row {
            border-bottom: 1px solid #e0e0e0;
            &:hover {
                background-color: #f5f5f5;
            }
            &-cell {
                padding: 10px 15px;
                text-align: left;
                border-bottom: 1px solid #e0e0e0;
            }
        }
    }
}