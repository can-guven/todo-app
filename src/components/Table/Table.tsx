import { FC } from "react";
import "./Table.scss";
interface Column {
  name: string;
  label: string;
  render?: (record: any) => JSX.Element;
}

interface TableProps {
  columns: Column[];
  dataSource: any[];
}

const Table: FC<TableProps> = (props) => {
  const { columns, dataSource } = props;

  return (
    <div>
      <h1>Table</h1>
      <table className="table">
        <thead className="table-head">
          <tr>
            {columns.map((column) => (
              <th key={column.name}>{column.label}</th>
            ))}
          </tr>
        </thead>
        <tbody className="table-body">
          {dataSource.map((record) => (
            <tr key={record.key}>
              {columns.map((column) => {
                if (column.render) {
                  return <td key={column.name}>{column.render(record)}</td>;
                }
                return <td key={column.name}>{record[column.name]}</td>;
              })}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default Table;
