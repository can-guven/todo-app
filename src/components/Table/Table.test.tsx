
import { render, screen } from "@testing-library/react";
import Table from "./Table";

describe('Table', () => {
    it('should render', () => {
        const { container } = render(<Table columns={[]} dataSource={[]} />);
        expect(container).toMatchSnapshot();
    });
    
    it('should render columns', () => {
        const columns = [
            {
                name: 'name',
                label: 'Name'
            },
            {
                name: 'priority',
                label: 'Priority'
            }
        ];
        const { container } = render(<Table columns={columns} dataSource={[]} />);
        expect(container).toMatchSnapshot();
    } );

    it('should render data', () => {
        const columns = [
            {
                name: 'name',
                label: 'Name'
            },
            {
                name: 'priority',
                label: 'Priority'
            }
        ];
        const dataSource = [
            {
                key: 1,
                name: 'name1',
                priority: 'priority1'
            },
            {
                key: 2,
                name: 'name2',
                priority: 'priority2'
            }
        ];
        const { container } = render(<Table columns={columns} dataSource={dataSource} />);
        expect(container).toMatchSnapshot();
    });

    it('should render data with render function', () => {
        const columns = [
            {
                name: 'name',
                label: 'Name'
            },
            {
                name: 'priority',
                label: 'Priority',
                render: (record: any) => <div>{record.priority}</div>
            }
        ];
        const dataSource = [
            {
                key: 2,
                name: 'name1',
                priority: 'priority1'
            },
            {
                key: 1,
                name: 'name2',
                priority: 'priority2'
            }
        ];
        const { container } = render(<Table columns={columns} dataSource={dataSource} />);
        expect(container).toMatchSnapshot();
    });

    it('should render data with render function and click', () => {
        const columns = [
            {
                name: 'name',
                label: 'Name'
            },
            {
                name: 'priority',
                label: 'Priority',
                render: (record: any) => <div>{record.priority}</div>
            }
        ];
        const dataSource = [
            {
                key : 1,
                name: 'name1',
                priority: 'priority1'
            },
            {
                key: 2,
                name: 'name2',
                priority: 'priority2'
            }
        ];
        const { container } = render(<Table columns={columns} dataSource={dataSource} />);
        expect(container).toMatchSnapshot();
        const div = screen.getByText('priority1');
        expect(div).toBeInTheDocument();
        expect(div).toHaveTextContent('priority1');
    });
    
});