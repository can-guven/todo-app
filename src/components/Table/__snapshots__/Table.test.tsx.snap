// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render 1`] = `
<div>
  <div>
    <h1>
      Table
    </h1>
    <table
      class="table"
    >
      <thead
        class="table-head"
      >
        <tr />
      </thead>
      <tbody
        class="table-body"
      />
    </table>
  </div>
</div>
`;

exports[`Table should render columns 1`] = `
<div>
  <div>
    <h1>
      Table
    </h1>
    <table
      class="table"
    >
      <thead
        class="table-head"
      >
        <tr>
          <th>
            Name
          </th>
          <th>
            Priority
          </th>
        </tr>
      </thead>
      <tbody
        class="table-body"
      />
    </table>
  </div>
</div>
`;

exports[`Table should render data 1`] = `
<div>
  <div>
    <h1>
      Table
    </h1>
    <table
      class="table"
    >
      <thead
        class="table-head"
      >
        <tr>
          <th>
            Name
          </th>
          <th>
            Priority
          </th>
        </tr>
      </thead>
      <tbody
        class="table-body"
      >
        <tr>
          <td>
            name1
          </td>
          <td>
            priority1
          </td>
        </tr>
        <tr>
          <td>
            name2
          </td>
          <td>
            priority2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should render data with render function 1`] = `
<div>
  <div>
    <h1>
      Table
    </h1>
    <table
      class="table"
    >
      <thead
        class="table-head"
      >
        <tr>
          <th>
            Name
          </th>
          <th>
            Priority
          </th>
        </tr>
      </thead>
      <tbody
        class="table-body"
      >
        <tr>
          <td>
            name1
          </td>
          <td>
            <div>
              priority1
            </div>
          </td>
        </tr>
        <tr>
          <td>
            name2
          </td>
          <td>
            <div>
              priority2
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should render data with render function and click 1`] = `
<div>
  <div>
    <h1>
      Table
    </h1>
    <table
      class="table"
    >
      <thead
        class="table-head"
      >
        <tr>
          <th>
            Name
          </th>
          <th>
            Priority
          </th>
        </tr>
      </thead>
      <tbody
        class="table-body"
      >
        <tr>
          <td>
            name1
          </td>
          <td>
            <div>
              priority1
            </div>
          </td>
        </tr>
        <tr>
          <td>
            name2
          </td>
          <td>
            <div>
              priority2
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
